os/arch/arm/src/armv7-a/arm_mmu.c

```diff
@@ -254,3 +289,264 @@ void mmu_invalidate_region(uint32_t vstart, size_t size)
+	// memset(addr, 0, L1_PGTBL_SIZE);
+	return addr;
+}
+
+/****************************************************************************
+ * Name: mmu_allocate_app_l2_pgtbl
+ *
+ * Description:
+ *   Allocate space for L2 page table of application, in accordance with
+ *   the requirements of the arch specific mmu.
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ * L2 Page table address
+ ****************************************************************************/
+uint32_t *mmu_allocate_app_l2_pgtbl(int app_id, int l2_idx)
+{
+	app_id--;
+	uint32_t *addr = (uint32_t *)(PGTABLE_BASE_VADDR + (2 * 16384) + (app_id * l2_idx * 1024));
+	// uint32_t *addr = (uint32_t *)kmm_memalign(L2_PGTBL_ALIGNMENT, L2_PGTBL_SIZE);
+	// ASSERT(addr);
+	// memset(addr, 0, L2_PGTBL_SIZE);
+	return addr;
+}
+
+/****************************************************************************
+ * Name: mmu_update_app_l1_pgtbl_ospgtbl
+ *
+ * Description:
+ * Loop through the L1 page table.
+ * Copy kernel L1 page table to app page table.
+ * If the entry is pointing to a L2 page table
+ * Allocate L2 page table for app.
+ * Copy entries from kernel to app L2 table.
+ * Update the L2 page table address in L1 table.
+ * 
+ * Input Parameters:
+ * app_pgtbl: Pointer to L1 page table of app
+ *
+ ****************************************************************************/
+void mmu_update_app_l1_pgtbl_ospgtbl(uint32_t *app_l1_pgtbl)
+{
+	uint32_t *os_l1_pgtbl = (uint32_t *)PGTABLE_BASE_VADDR;
+
+	memcpy((void *)app_l1_pgtbl, (void *)os_l1_pgtbl, L1_PGTBL_SIZE);
+	cp15_flush_dcache((uintptr_t)app_l1_pgtbl, (uintptr_t)app_l1_pgtbl + L1_PGTBL_SIZE);
+
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+	for (int i = 0; i < L1_PGTBL_NENTRIES; i++) {
+		if ((os_l1_pgtbl[i] & PMD_TYPE_MASK) == PMD_TYPE_PTE) {
+			//Found a L2 page table.
+			uint32_t *os_l2_pgtbl = (uint32_t *)(os_l1_pgtbl[i] & PMD_PTE_PADDR_MASK);
+			uint32_t *app_l2_pgtbl = mmu_allocate_app_l2_pgtbl();
+			memcpy(app_l2_pgtbl, os_l2_pgtbl, L2_PGTBL_SIZE);
+			app_l1_pgtbl[i] &= ~PMD_PTE_PADDR_MASK;
+			app_l1_pgtbl[i] |= (uint32_t)app_l2_pgtbl & PMD_PTE_PADDR_MASK;
+		}	
+	}
+#endif
+}
+
+/****************************************************************************
+ * Name: mmu_map_app_region
+ *
+ * Description
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ ****************************************************************************/
+void mmu_map_app_region(int app_id, uint32_t *l1_pgtbl, uint32_t start, uint32_t size, bool ro, bool exec, bool global)
+{
+	uint32_t idx;
+	uint32_t val;
+	uint32_t end = start + size;
+	irqstate_t flags;
+
+	lldbg("start = 0x%08x end = 0x%08x size = %x\n", start, end, size);
+
+	// Run a loop until the entire region is mapped.
+	while (start < end) {
+		// Check if this address can be mapped to a section.
+		if (!(start & SECTION_MASK) && !(size & SECTION_MASK)) {
+			// Yes. Update the section entry in the the L1 page table.
+			idx = start >> 20;
+			val = start & PMD_PTE_PADDR_MASK;
```

Comment:
Not required at present, but may be needed to change the mask to section instead of PTE